enhancements/installer/azure-internal-with-user-defined-routing.md

```diff
@@ -0,0 +1,243 @@
+---
+title: azure-internal-with-user-defined-outbound-routing
+authors:
+  - "@abhinavdahiya"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-20
+last-updated: 2020-05-20
+status: implementable
+see-also:
+  - "./enhancements/azure-private-internal-clusters.md"
+  - https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype
+  - https://github.com/openshift/installer/pull/3324
+replaces:
+superseded-by:
+---
+
+# Azure Internal Clusters with User Defined Outbound Routing
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently for internal clusters on Azure, OpenShift clusters always use Public Standard Loadbalancers for Internet egress. This configuration requires public IPs, public Loadbalancers etc. which some customers do not want for their internal clusters. This proposal allows the users to choose their own outbound routing for Internet allowing them to use any pre-existing setups instead of the per-cluster OpenShift recommended way.
+
+## Motivation
+
+Internal clusters on Azure require explicit setup to allow outbound routing or egress to Internet to pull container images, updates, access cloud APIs etc. Currently the cluster uses a public standard loadbalancer, with only outbound_rules, to provide the outbound routing to Internet. But this force customers that either have pre-existing setups like PROXY for outbound routing or do not require any egress to Internet at all i.e. air-gapped, into the above presrcribed setup without allowing them to change it. And often for customers that are interested in internal clsuters already have policies that disallow creation resources like public IP addresses, which causes the installations to fail and forces the customers to use user-provided workflow.
```

Comment:
I think the OpenShift installer's definition of "outbound rules" differs from [Azure's definition](https://docs.microsoft.com/en-gb/azure/load-balancer/load-balancer-outbound-rules-overview), and now would be a good time to align the two (ARO has already done this).  In "private API server" mode, I believe the installer sets a dummy inbound rule to connect the public LB with the API servers, calling it an outbound rule.  In fact I believe it should create an Azure outbound rule.  Doing this has a few advantages:
* alignment with Azure best practices and ARO
* simplifies installer, ARO and customer logic around whether to use disableOutboundSNAT
* brings us closer to being able to enable stable IP addresses for egress traffic when LB is used, which is a separate customer requirement